# Data Consistency Checks

## Table of Contents
#### 1. Importing libraries and Data
#### 2. Data Consistency Checks on Products data
##### 2.1 Mixed Data Types
##### 2.2 Missing values
##### 2.3 Duplicates
#### 3. Data Consistency Checks on Orders data
##### 3.1 Mixed Data Types
##### 3.2 Missing values
##### 3.3 Duplicates

## 01. Importing libraries and Data

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Main path
path = r'/Users/helenarobalinho/Desktop/Data Analytics/2. Data Immersion/Achievement 4 - Python Fundamentals for Data Analysts/03-2026 Instacart Basket Analysis'

In [4]:
# Import products.csv from Original Data folder
df_products = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import orders.csv from Prepared Data folder
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## 02. Data Consistency Checks on Products data

In [6]:
df_products.describe()

,product_id,aisle_id,department_id,prices
count,49693.000000,49693.000000,49693.000000,49693.000000
mean,24844.345139,67.770249,11.728433,9.994136
std,14343.717401,38.316774,5.850282,453.519686
min,1.000000,1.000000,1.000000,1.000000
25%,12423.000000,35.000000,7.000000,4.100000
50%,24845.000000,69.000000,13.000000,7.100000
75%,37265.000000,100.000000,17.000000,11.200000
max,49688.000000,134.000000,21.000000,99999.000000


### 2.1 Mixed Data Types

In [8]:
# Create a dataframe (for simulation purposes)
df_test = pd.DataFrame ()

In [9]:
# Create a mixed type column
df_test ['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

,mix
0,a
1,b
2,1
3,True


In [11]:
# Function to check if the df contains mixed-type columns (applymap foi substituido por map)

for col in df_test.columns.tolist ():
    weird = (df_test [[col]].map(type) != df_test [[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test [weird]) > 0:
        print (col)

mix


In [12]:
# Fix the mixed data types by forcing ot to be all string
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
df_test['mix'].head()

0       a
1       b
2       1
3    True
Name: mix, dtype: object

### 2.2 Missing Values

In [14]:
# Finding the missing values
df_products.isnull(). sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# Viewing the missing data
df_nan = df_products[df_products ['product_name'].isnull() == True]

In [16]:
df_nan

,product_id,product_name,aisle_id,department_id,prices
33,34,NaN,121,14,12.2
68,69,NaN,26,7,11.8
115,116,NaN,93,3,10.8
261,262,NaN,110,13,12.1
525,525,NaN,109,11,1.2
1511,1511,NaN,84,16,14.3
1780,1780,NaN,126,11,12.3
2240,2240,NaN,52,1,14.2
2586,2586,NaN,104,13,12.4
3159,3159,NaN,126,11,13.1


- The missing values are small and cannot be guessed, the product_name is a string - Let's drop it.

In [17]:
# Since overwriting is risky, we can create a subset dataframe:

df_products_clean = df_products[df_products['product_name'].isnull() == False]

In [18]:
# Let's compare. Before:
df_products.shape

(49693, 5)

In [19]:
# After:
df_products_clean.shape

(49677, 5)

### Exactly 16 rows were dropped from the original df ✅ 

### 2.3 Duplicates

In [20]:
# Creating a subset of the duplicate rows to find them:
df_dups = df_products_clean[df_products_clean.duplicated()]

In [21]:
df_dups

,product_id,product_name,aisle_id,department_id,prices
462,462,Fiber 4g Gummy Dietary Supplement,70,11,4.8
18459,18458,Ranger IPA,27,5,9.2
26810,26808,Black House Coffee Roasty Stout Beer,27,5,13.4
35309,35306,Gluten Free Organic Peanut Butter & Chocolate ...,121,14,6.8
35495,35491,Adore Forever Body Wash,127,11,9.9


In [22]:
# Removing the duplicates by creating a new version of the df without them:
df_products_clean_no_dups = df_products_clean.drop_duplicates()

In [23]:
# Let's compare it. Before:
df_products_clean.shape

(49677, 5)

In [24]:
# After:
df_products_clean_no_dups.shape

(49672, 5)

### Exactly 5 duplicates were removed ✅

## 3. Data Consistency Checks on Orders data

Run the df. describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [25]:
df_orders.describe()

,Unnamed: 0,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710541e+06,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552705e+05,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710541e+06,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421082e+06,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


- The maximum order_number is 100. The count of the days_since_prior_order is lower compared to the other columns - there could be missing values in this column.

### 3.1 Mixed Data Types

Check for mixed-type data in your df_ords dataframe

In [26]:
# Function to check if the df contains mixed-type columns (applymap foi substituido por map)

for col in df_orders.columns.tolist ():
    weird = (df_orders [[col]].map(type) != df_orders [[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_orders [weird]) > 0:
        print (col)

- There are no columns with mixed data type in the dataframe df_orders

In [27]:
df_orders.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### 3.2. Missing Values
Run a check for missing values in your df_ords dataframe.

In [28]:
# finding missing values:
df_orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

- There are 206209 missing values in the column days_since_prior_order. The missing values could be for the customers during their first order. Values cannot be imputed, and cannot be removed. Maybe we can introduce a new flag for 1st order.

##### Addressing the missing values using an appropriate method:

##### Flagging

##### Before addressing the missing values, we need to check if the records with the missing values are having value 1 in the column 'order_number'

In [29]:
# This created a df with just the NaN rows:
df_orders_nan = df_orders[df_orders['days_since_prior_order'].isnull() == True]

In [30]:
# Let's view the missing data:
df_orders_nan

,Unnamed: 0,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order
0,0,2539329,1,prior,1,2,8,NaN
11,11,2168274,2,prior,1,2,11,NaN
26,26,1374495,3,prior,1,1,14,NaN
39,39,3343014,4,prior,1,6,11,NaN
45,45,2717275,5,prior,1,3,12,NaN
...,...,...,...,...,...,...,...,...
3420930,3420930,969311,206205,prior,1,4,12,NaN
3420934,3420934,3189322,206206,prior,1,3,18,NaN
3421002,3421002,2166133,206207,prior,1,6,19,NaN
3421019,3421019,2227043,206208,prior,1,1,15,NaN


- Yes, all the NaN are having 1 in the 'order_number' column.
- This shows that all records with missing values in the column days_since_prior_order are customer's first orders. So these records cannot be deleted. Hence we can introduce a new flag to the dataframe.

In [31]:
#  Flagging
df_orders_first_order = df_orders[df_orders['order_number'] == 1]

In [32]:
df_orders_first_order

,Unnamed: 0,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order
0,0,2539329,1,prior,1,2,8,NaN
11,11,2168274,2,prior,1,2,11,NaN
26,26,1374495,3,prior,1,1,14,NaN
39,39,3343014,4,prior,1,6,11,NaN
45,45,2717275,5,prior,1,3,12,NaN
...,...,...,...,...,...,...,...,...
3420930,3420930,969311,206205,prior,1,4,12,NaN
3420934,3420934,3189322,206206,prior,1,3,18,NaN
3421002,3421002,2166133,206207,prior,1,6,19,NaN
3421019,3421019,2227043,206208,prior,1,1,15,NaN


- All first orders are marked with the flag.

### 3.3 Duplicates
Run a check for duplicate values in your df_ords data

In [33]:
# Subset of the duplicate rows:
df_orders_dups = df_orders[df_orders.duplicated()]

In [34]:
df_orders_dups

,Unnamed: 0,order_id,user_id,eval_set,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order


- There are no duplicated records in the orders Dataframe

## Exporting the final, cleaned df_products and df_orders data as .csv files in the Prepared Data folder

In [35]:
# Export the cleaned dataframe for df_products
df_products_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [36]:
# Export the cleaned dataframe for df_orders
df_orders.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [38]:
df_orders.shape

(3421083, 8)